waddrmgr/db.go

```diff
@@ -1738,212 +2154,115 @@ func upgradeManager(db walletdb.DB, namespaceKey []byte, pubPassPhrase []byte,
-		}
+	if err := mainBucket.Delete(coinTypePrivKeyName); err != nil {
+		return err
+	}
+	if err := mainBucket.Delete(coinTypePubKeyName); err != nil {
+		return err
+	}
 
-		var oldName string
+	// Next, we'll move over everything that was in the meta bucket to the
+	// meta bucket within the new scope.
+	metaBucket := ns.NestedReadWriteBucket(metaBucketName)
+	lastAccount := metaBucket.Get(lastAccountName)
+	if err := metaBucket.Delete(lastAccountName); err != nil {
+		return err
+	}
 
-		// Delete any other names for the default account.
-		c := ns.NestedReadWriteBucket(acctNameIdxBucketName).ReadWriteCursor()
-		for k, v := c.First(); k != nil; k, v = c.Next() {
-			// Skip nested buckets.
-			if v == nil {
-				continue
-			}
+	scopedMetaBucket := bip44Bucket.NestedReadWriteBucket(metaBucketName)
+	err = scopedMetaBucket.Put(lastAccountName, lastAccount)
+	if err != nil {
+		return err
+	}
 
-			// Skip account names which aren't for the default account.
-			account := binary.LittleEndian.Uint32(v)
-			if account != DefaultAccountNum {
-				continue
+	// Finally, we'll mechanically move over a set of keys which were
+	// formerly under the main bucket, into the new scoped buckets. We'll
+	// do so mechanically by obtaining a slice of all the keys that we need
+	// to modify.
+	keysToMigrate := [][]byte{
+		acctBucketName, addrBucketName, usedAddrBucketName,
+		addrAcctIdxBucketName, acctNameIdxBucketName, acctIDIdxBucketName,
+	}
+	for _, bucketKey := range keysToMigrate {
+		// Within this bucket key, we'll migrate over, then delete each
+		// key.
+		bucketToMigrate := mainBucket.NestedReadWriteBucket(bucketKey)
+		err := bucketToMigrate.ForEach(func(k, v []byte) error {
+			if err := bip44Bucket.Put(k, v); err != nil {
+				return err
 			}
 
-			if !bytes.Equal(k[4:], []byte(acctInfo.name)) {
-				err := c.Delete()
-				if err != nil {
-					const str = "error deleting default account alias"
-					return managerError(ErrUpgrade, str, err)
-				}
-				oldName = string(k[4:])
-				break
+			if err := mainBucket.Delete(k); err != nil {
+				return err
 			}
-		}
-
-		// The account number to name index may map to the wrong name,
-		// so rewrite the entry with the true name from the account row
-		// instead of leaving it set to an incorrect alias.
-		err = putAccountIDIndex(ns, DefaultAccountNum, acctInfo.name)
-		if err != nil {
-			const str = "account number to name index could not be " +
-				"rewritten with actual account name"
-			return managerError(ErrUpgrade, str, err)
-		}
-
-		// Ensure that the true name for the default account maps
-		// forwards and backwards to the default account number.
-		name, err := fetchAccountName(ns, DefaultAccountNum)
-		if err != nil {
 			return err
-		}
-		if name != acctInfo.name {
-			const str = "account name index does not map default account number to correct name"
-			return managerError(ErrUpgrade, str, nil)
-		}
-		acct, err := fetchAccountByName(ns, acctInfo.name)
+		})
```

Comment:
When upgrading, this throws a nil dereference as follows:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x40 pc=0x7bec68]

goroutine 1 [running]:
github.com/roasbeef/btcwallet/waddrmgr.upgradeToVersion5(0xfe1960, 0xc42001a8c0, 0xc4201268f0, 0x6, 0x8, 0xc42001a8c0, 0x0)
        /home/alex/go/src/github.com/roasbeef/btcwallet/waddrmgr/db.go:2256 +0xaa8
github.com/roasbeef/btcwallet/waddrmgr.upgradeManager.func2(0xfe0580, 0xc420128050, 0xc420128050, 0x0)
        /home/alex/go/src/github.com/roasbeef/btcwallet/waddrmgr/db.go:2134 +0x8d
github.com/roasbeef/btcwallet/walletdb.Update(0xfde860, 0xc42013e3c0, 0xc420125b78, 0x0, 0x0)
        /home/alex/go/src/github.com/roasbeef/btcwallet/walletdb/interface.go:218 +0x72
github.com/roasbeef/btcwallet/waddrmgr.upgradeManager(0xfde860, 0xc42013e3c0, 0xfd3708, 0x8, 0x8, 0xc4201268f0, 0x6, 0x8, 0x1016c60, 0xc420125e20, ...)
        /home/alex/go/src/github.com/roasbeef/btcwallet/waddrmgr/db.go:2135 +0x450
github.com/roasbeef/btcwallet/waddrmgr.DoUpgrades(0xfde860, 0xc42013e3c0, 0xfd3708, 0x8, 0x8, 0xc4201268f0, 0x6, 0x8, 0x1016c60, 0xc420125e20, ...)
        /home/alex/go/src/github.com/roasbeef/btcwallet/waddrmgr/manager.go:1498 +0xa1
github.com/roasbeef/btcwallet/wallet.Open(0xfde860, 0xc42013e3c0, 0xc4201268f0, 0x6, 0x8, 0xc420125e20, 0x1016c60, 0x0, 0x0, 0x0)
        /home/alex/go/src/github.com/roasbeef/btcwallet/wallet/wallet.go:2692 +0xf2
github.com/roasbeef/btcwallet/wallet.(*Loader).OpenExistingWallet(0xc42019a870, 0xc4201268f0, 0x6, 0x8, 0x1, 0x0, 0x0, 0x0)
        /home/alex/go/src/github.com/roasbeef/btcwallet/wallet/loader.go:178 +0x3c6
main.walletMain(0x0, 0x0)
        /home/alex/go/src/github.com/roasbeef/btcwallet/btcwallet.go:96 +0x6d0
main.main()
        /home/alex/go/src/github.com/roasbeef/btcwallet/btcwallet.go:33 +0x32
```

Reply:
Thanks for the fix! Pushed a version with that applied.